Sources/TuistKit/Generator/ProjectGroups.swift

```diff
@@ -1,34 +1,59 @@
 import Basic
 import Foundation
+import TuistCore
 import xcodeproj
 
+enum ProjectGroupsError: FatalError, Equatable {
+    case missingGroup(String)
+
+    var description: String {
+        switch self {
+        case let .missingGroup(group):
+            return "Couldn't find group: \(group)"
+        }
+    }
+
+    var type: ErrorType {
+        switch self {
+        case .missingGroup:
+            return .bug
+        }
+    }
+}
+
 class ProjectGroups {
     // MARK: - Attributes
 
     let main: PBXGroup
     let products: PBXGroup
     let projectManifest: PBXGroup
-    let project: PBXGroup
     let frameworks: PBXGroup
     let playgrounds: PBXGroup?
-    let pbxproj: PBXProj
+
+    private let pbxproj: PBXProj
+    private let projectGroups: [String: PBXGroup]
+    private let knownGroups: OrderedSet<PBXGroup>
 
     // MARK: - Init
 
-    init(main: PBXGroup,
-         products: PBXGroup,
-         projectManifest: PBXGroup,
-         frameworks: PBXGroup,
-         project: PBXGroup,
-         playgrounds: PBXGroup?,
-         pbxproj: PBXProj) {
+    private init(main: PBXGroup,
+                 projectGroups: [(name: String, group: PBXGroup)],
+                 products: PBXGroup,
+                 projectManifest: PBXGroup,
+                 frameworks: PBXGroup,
+                 playgrounds: PBXGroup?,
+                 pbxproj: PBXProj) {
         self.main = main
+        self.projectGroups = Dictionary(uniqueKeysWithValues: projectGroups)
         self.products = products
         self.projectManifest = projectManifest
         self.frameworks = frameworks
-        self.project = project
         self.playgrounds = playgrounds
         self.pbxproj = pbxproj
+
+        let allProjectGroups = projectGroups.map { $0.group }
+        let predefinedGroups = [projectManifest, frameworks, playgrounds, products].compactMap { $0 }
+        knownGroups = OrderedSet(allProjectGroups + predefinedGroups)
```

Comment:
Considering that all the attributes on this class are immutable. We can make `knownGroups` a lazy variable and move all this logic into the variable initializer. What do you think?